### PR TITLE
Ecdc 3350 timepix 2

### PR DESCRIPTION
**Pull request by JennyLouise**
### Issue reference / description

The branch you merge from should already reference an event-formation-unit github ticket number. You can add a descriptive title, but if an issue is referenced, you don't have to.

## Checklist for submitter

- [ ] Check for conflict with integration test
- [ ] Unit tests pass

---

### Nominate for Group Code Review

- [ ] Nominate for code review

### CODE REVIEW

File: src/common/reduction/clustering/Hierarchical2DClusterer.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    visited[i] = true;
+    for (uint j = i + 1; j < clusterSize; j++) {
+      if (visited[j]) {
+        continue; // skip points that have already been visited
+      }
+      double x_distance = (double)current_time_cluster_[i].x_coordinate -
+                          (double)current_time_cluster_[j].x_coordinate;
+      double y_distance = (double)current_time_cluster_[i].y_coordinate -
+                          (double)current_time_cluster_[j].y_coordinate;
+      double distance = sqrt(pow(x_distance, 2) + pow(y_distance, 2));
```
Comment (JennyLouise) [inline src/common/reduction/clustering/Hierarchical2DClusterer.cpp:68]: Alternatively could not square root this, but square the threshold value instead - improve performance

File: src/modules/caen/test/CaenBaseTest.cpp [hunk 2]
```diff
@@ -250,10 +246,11 @@ TEST_F(CaenBaseTest, DataReceiveGoodMiracles) {
   Settings.ConfigFile = MIRACLES_CONFIG;
   Settings.CalibFile = MIRACLES_CALIB;
 
-  Settings.DetectorPort = 9000;
+  Settings.DetectorPort = 9216;
   Settings.UpdateIntervalSec = 0;
   Settings.DumpFilePrefix = "deleteme_";
-  CaenBaseStandIn Readout(Settings, ESSReadout::Parser::MIRACLES);
+  Caen::CaenBase Readout(Settings, ESSReadout::Parser::MIRACLES);
+  Readout.Counters = {};
```
Comment (JennyLouise) [inline src/modules/caen/test/CaenBaseTest.cpp:253]: Probably don't need this line now

File: src/modules/timepix3/Timepix3Instrument.h [hunk 3]
```diff
@@ -13,14 +13,23 @@
...
+#include <timepix3/geometry/Timepix3Geometry.h>
 
 namespace Timepix3 {
 
+struct TimepixHit {
+  uint32_t X;
+  uint32_t Y;
+  uint64_t TimeOfFlight;
+  uint16_t ToT;
+};
```
Comment (JennyLouise) [inline src/modules/timepix3/Timepix3Instrument.h:31]: I don't think this is necessary

File: src/modules/caen/test/CaenBaseTest.cpp (hunk 2 above)
Comment (tbukovics) [reply src/modules/caen/test/CaenBaseTest.cpp]: Agree, not necessary, I will remove this line

File: src/modules/timepix3/Timepix3Instrument.h (hunk 3 above)
Comment (tbukovics) [reply src/modules/timepix3/Timepix3Instrument.h:31]: Indeed, there are no references in the source code to this object anymore. I will remove

File: src/common/reduction/clustering/Hierarchical2DClusterer.cpp (hunk 1 above)
Comment (tbukovics) [reply src/common/reduction/clustering/Hierarchical2DClusterer.cpp]: I reworked this part.

File: src/common/reduction/clustering/Hierarchical2DClusterer.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      if (visited[j]) {
+        continue; // skip points that have already been visited
+      }
+      double x_distance = (double)current_time_cluster_[i].x_coordinate -
+                          (double)current_time_cluster_[j].x_coordinate;
+      double y_distance = (double)current_time_cluster_[i].y_coordinate -
+                          (double)current_time_cluster_[j].y_coordinate;
+
+      // Calculate distance according to d2 = dx2 + dy2 to remove sqrt calculation
+      double distance_sqr = sqrt(x_distance) + sqrt(y_distance);
```
Comment (tbukovics) [inline src/common/reduction/clustering/Hierarchical2DClusterer.cpp:71]: @mortenjc could you have a look on this change. I added an inline sqr function and also compare sqr(threshold) with sqr(distance). Now pow() and sqrt() eliminated.

File: src/common/reduction/clustering/Hierarchical2DClusterer.cpp [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      if (visited[j]) {
+        continue; // skip points that have already been visited
+      }
+      double x_distance = (double)current_time_cluster_[i].x_coordinate -
+                          (double)current_time_cluster_[j].x_coordinate;
+      double y_distance = (double)current_time_cluster_[i].y_coordinate -
+                          (double)current_time_cluster_[j].y_coordinate;
+
+      // Calculate distance according to d2 = dx2 + dy2 to remove sqrt calculation
+      double distance_sqr = sqr(x_distance) + sqr(y_distance);
```
Comment (tbukovics) [inline src/common/reduction/clustering/Hierarchical2DClusterer.cpp:71]: Sorry for the spam, but I realized that I made a typo, so I corrected it. 
@mortenjc could you have a look on this change. I added an inline sqr function and also compare sqr(threshold) with sqr(distance). Now pow() and sqrt() eliminated.

**Review by mortenjc — APPROVED** [review]

File: src/common/reduction/clustering/Hierarchical2DClusterer.cpp [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    for (uint j = i + 1; j < clusterSize; j++) {
+      if (visited[j]) {
+        continue; // skip points that have already been visited
+      }
+      double x_distance = (double)current_time_cluster_[i].x_coordinate -
+                          (double)current_time_cluster_[j].x_coordinate;
+      double y_distance = (double)current_time_cluster_[i].y_coordinate -
+                          (double)current_time_cluster_[j].y_coordinate;
+
+      // Calculate distance according to d2 = dx2 + dy2 to remove sqrt calculation
```
[the remark this answers is not in the document]
Comment (tbukovics) [reply src/common/reduction/clustering/Hierarchical2DClusterer.cpp:70]: Tracing the code x/y_distance and also the distance_sqr is local variable and only used at line 80 to compare with the gap. I also see that it's used in XTRACE messages which I'm updated:
```c++
XTRACE(DATA, DEB,
             "Determined squere of the distance between points is %f, the squere of threshold is %u",
             distance_sqr, max_coord_gap_sqr_);
```
I will apply your suggestion on the comment.

**Pull request merged**